### PR TITLE
FEA-2770 FEA-2772 Action v2 type migration

### PR DESCRIPTION
**Pull request by sorenthompson-wk**
## Motivation
W_flux Action to ActionV2 codemod

## Changes
make codemod to migrate Action to ActionV2

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Architecture member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/w_flux/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/w_flux/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: w_flux_codemod/lib/src/action_v2_suggestor.dart [hunk 1]
```diff
@@ -51,48 +45,60 @@ mixin ActionV2Migrator on AstVisitingSuggestor {
...
   @override
-  bool shouldMigrate(InvocationExpression node) =>
-      node.staticInvokeType != null &&
-      node.argumentList.arguments.isNotEmpty &&
-      node.function.staticParameterElement.name == 'dispatch';
+  bool shouldMigrate(node) {
+    if (node is FunctionExpressionInvocation) {
+      final name = node.function.staticType?.element?.displayName;
+      // the type migration should have happened prior to this suggestor.
+      return name == 'ActionV2';
```
Comment (sorenthompson-wk) [inline w_flux_codemod/lib/src/action_v2_suggestor.dart:52]: Should we move this suggestor to an action_v2_migrate_step_3? Or does placing it at the end of list enough to ensure that it will run last?

File: w_flux_codemod/test/action_v2_suggestor_test.dart [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+          'var action; action = Action();',
+          'var action; action = ActionV2();',
+        );
+        testSuggestor(
+          'field initialization',
+          suggestor,
+          'class C { var action; C() { action = Action(); } }',
+          'class C { var action; C() { action = ActionV2(); } }',
+        );
+        testSuggestor(
```
Comment (sorenthompson-wk) [inline w_flux_codemod/test/action_v2_suggestor_test.dart:135]: @evanweible-wf what was the original intention with this test? Should the suggestor actually be skipping the dynamic functions?

File: w_flux_codemod/lib/src/action_v2_suggestor.dart (hunk 1 above)
Comment (evanweible-wf) [reply w_flux_codemod/lib/src/action_v2_suggestor.dart:52]: Hmm.. well I guess my understanding of how `runInteractiveCodemodSequence` is wrong (or there's a bug), because I tried this and it didn't work. Had to separate them out to two separate calls to `runInteractiveCodemod` like so:

```dart
  exitCode = await runInteractiveCodemod(
    filePathsFromGlob(Glob('**.dart', recursive: true)),
    aggregate([
      ActionV2FieldAndVariableMigrator(),
      ActionV2ParameterMigrator(),
      ActionV2ReturnTypeMigrator(),
      ActionV2SuperTypeMigrator(),
    ]),
    args: args,
  );
  if (exitCode != 0) {
    return;
  }
  exitCode = await runInteractiveCodemod(
    filePathsFromGlob(Glob('**.dart', recursive: true)),
    ActionV2DispatchMigrator(),
    args: args,
  );
```

File: w_flux_codemod/test/action_v2_suggestor_test.dart
Comment (evanweible-wf) [inline w_flux_codemod/test/action_v2_suggestor_test.dart:1]: I think the only test coverage we're missing here are test cases for each suggestor to verify that it handles the w_flux import being prefixed and that it ignores same-named types that aren't actually from w_flux. I'll include an example of this in a PR

File: w_flux_codemod/README.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    - step by step:
+    ```bash
+    dart pub global run w_flux_codemod:action_v2_migrate_step_1.dart
+    dart pub global run w_flux_codemod:action_v2_migrate_step_2.dart
+    ```
+
+    - all at once:
+    ```bash
+    dart pub global run w_flux_codemod:action_v2_migrate.dart
```
Comment (evanweible-wf) [inline w_flux_codemod/README.md:48]: Shouldn't need the `.dart` here since we're defining names for these executables in the pubspec:
```suggestion
    dart pub global run w_flux_codemod:action_v2_migrate_step_1
    dart pub global run w_flux_codemod:action_v2_migrate_step_2
    ```

    - all at once:
    ```bash
    dart pub global run w_flux_codemod:action_v2_migrate
```

File: w_flux_codemod/lib/src/action_v2_suggestor.dart [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+/// TODO - will likely require overriding [visitInvocationExpression]
+/// and checking the type of the thing being invoked to see if it's an Action
+/// If it is and there are no arguments passed, need to pass in `null`
+class ActionV2DispatchMigrator extends RecursiveAstVisitor
+    with AstVisitingSuggestor, ActionV2Migrator {
+  @override
+  bool shouldMigrate(node) {
+    if (node is FunctionExpressionInvocation) {
+      final name = node.function.staticType?.element?.displayName;
```
[the remark this answers is not in the document]
Comment (sorenthompson-wk) [reply w_flux_codemod/lib/src/action_v2_suggestor.dart:50]: name was deprecated or at least that's what the analyzer was saying.

[the remark this answers is not in the document]
Comment (evanweible-wf) [reply w_flux_codemod/lib/src/action_v2_suggestor.dart:50]: Huh that's weird, I'm not seeing it as deprecated. Maybe that was a different `name` field on a different node type?

[the remark this answers is not in the document]
Comment (sorenthompson-wk) [reply w_flux_codemod/lib/src/action_v2_suggestor.dart:50]: that's weird I'm seeing it as good now!

**Review by evanweible-wf — DISMISSED** [review]
This is looking good to me! Let's run it in a few more repos to see if we can catch any other edge cases. Once we feel good about that, we can get this merged and released.

File: w_flux_codemod/test/action_v2_suggestor_test.dart [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+        importMode: WFluxImportMode.multipleShown,
+      );
+
+      testSuggestor(
+        'hidden import',
+        suggestor,
+        '',
+        '',
+        importMode: WFluxImportMode.hidden,
+      );
```
[the remark this answers is not in the document]
Comment (sorenthompson-wk) [reply w_flux_codemod/test/action_v2_suggestor_test.dart:112]: These are just test the import statements, and those are already added, so we don't need any additional code.

[the remark this answers is not in the document]
Comment (evanweible-wf) [reply w_flux_codemod/test/action_v2_suggestor_test.dart:112]: Oops sorry, think this was leftover from a previous review

**Review by evanweible-wf — APPROVED** [review]

**Comment by evanweible-wf** [comment]
LGTM! I ran this in wdesk_sdk and graph_app again and it works great! Had to make two manual changes in wdesk_sdk, but none in graph_app. Let's ship it!

**Comment by evanweible-wf** [comment]
QA +1

**Comment by evanweible-wf** [comment]
@Workiva/release-management-p

**Comment by sorenthompson-wk** [comment]
@Workiva/release-management-p send it

**Review by kirkrichardson-wk — APPROVED** [review]

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM
